Support DDL On Cluster to stop waiting for offline hosts

tuanpach opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
1. Refactor `DDLQueryStatusSource`:
  * Rename `DDLQueryStatusSource` to `DistributedQueryStatusSource`, and make it a base class
  * Create two subclasses `DDLOnClusterQueryStatusSource` and `ReplicatedDatabaseQueryStatusSource` derived from `DDLQueryStatusSource` to query the status of DDL tasks from `DDL On Cluster and Replicated databases respectively.
2. Support stop waiting for offline hosts in `DDLOnClusterQueryStatusSource`

Closes #66013

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

tuanpach committed:
1) Create replica_dir in DDLWorker
2) Mark replicas active in DDLWorker

tuanpach committed:
Merge branch 'report-running-ddl-worker-hosts' into support-ddl-on-cluster-stop-waiting-offline-hosts

tuanpach committed:
1) Refactor DDLQueryStatusSource:
- Rename DDLQueryStatusSource to DistributedQueryStatusSource, and make it as a base class
- Create two subclasses: DDLOnClusterQueryStatusSource, ReplicatedDatabaseQueryStatusSource derived from DDLQueryStatusSource
2) Support stop waiting offline hosts in DDLOnClusterQueryStatusSource

robot-clickhouse commented:
<!-- automatic status comment for PR #69660 from tuanpach/ClickHouse:support-ddl-on-cluster-stop-waiting-offline-hosts -->
*This is an automated comment for commit 4be513451800d54c611f6f3211797cbc40717320 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69660/4be513451800d54c611f6f3211797cbc40717320/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/7b94dc1813ae2936fd0e276eb46b45ef6fcae994/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

tuanpach committed:
Add test_ddl_worker_replicas

tuanpach committed:
Add test_ddl_on_cluster_stop_waiting_for_offline_hosts test

tuanpach committed:
Fix compilation error

tuanpach committed:
Fix fillHostStatus in ReplicatedDatabaseQueryStatusSource

tuanpach committed:
No need to  and  in DatabaseReplicated

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts

tuanpach committed:
Merge branch 'report-running-ddl-worker-hosts' into support-ddl-on-cluster-stop-waiting-offline-hosts

tuanpach committed:
Add distributed_ddl.replicas_path setting

tuanpach committed:
Merge branch 'report-running-ddl-worker-hosts' into support-ddl-on-cluster-stop-waiting-offline-hosts

tuanpach committed:
Pass repliaces_path to DistributedQueryStatusSource

tuanpach committed:
Remove sleep in test_ddl_worker_replicas test

tuanpach committed:
Reformat the test

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts

tuanpach committed:
Not re-create an ephemeral node if it exists

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts

tuanpach committed:
Fix style check error

tavplubix commented on tests/integration/test_ddl_on_cluster_stop_waiting_for_offline_hosts/test.py:
```diff
@@ -0,0 +1,90 @@
+    timeout = 10
+    settings = {"distributed_ddl_task_timeout": timeout}
+
+    start = time.time()
+    node1.query(
+        "DROP TABLE IF EXISTS test_table ON CLUSTER test_cluster SYNC",
+        settings=settings,
+    )
+    assert time.time() - start < timeout
```
The query would throw an exception in case of timeout, so looks like the assert is redundant

tavplubix commented on tests/integration/test_ddl_on_cluster_stop_waiting_for_offline_hosts/test.py:
```diff
@@ -0,0 +1,90 @@
+    node4.stop()
+
+    start = time.time()
+    with pytest.raises(Exception) as err:
+        node1.query(
+            "DROP TABLE IF EXISTS test_table ON CLUSTER test_cluster SYNC",
+            settings=settings,
+        )
+    assert "Return code: 159" in str(err.value)
```
JFYI you can use smth like `assert "<error message>" in node1.query_and_get_error("<query>")`

tuanpach committed:
Tidy the test file

tuanpach committed:
1) Extend stop_start_wait_sec and reduce connection_timeout_ms in
test_replicated_database when restarting an instance.
2) Make createReplicaDirs and markReplicasActive virtual functions

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts

tuanpach committed:
Fix flaky test

tuanpach committed:
Revert changes in test_replicated_database

tuanpach committed:
Merge branch 'report-running-ddl-worker-hosts' into support-ddl-on-cluster-stop-waiting-offline-hosts

tuanpach committed:
Tidy test_ddl_on_cluster_stop_waiting_for_offline_hosts

tuanpach committed:
Make test_ddl_worker_replicas predictable

tuanpach committed:
Merge branch 'report-running-ddl-worker-hosts' into support-ddl-on-cluster-stop-waiting-offline-hosts

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts

tuanpach committed:
Merge branch 'report-running-ddl-worker-hosts' into support-ddl-on-cluster-stop-waiting-offline-hosts

tuanpach committed:
Use query_id to generate tempoerate table name when replacing tables

tuanpach committed:
Merge remote-tracking branch 'origin/master' into support-ddl-on-cluster-stop-waiting-offline-hosts

tuanpach committed:
Merge remote-tracking branch 'origin/master' into support-ddl-on-cluster-stop-waiting-offline-hosts

tavplubix commented:
Stateless tests - 00653_verification_monotonic_data_load - https://github.com/ClickHouse/ClickHouse/pull/68614#issuecomment-2441611537

evillique commented on tests/integration/helpers/cluster.py:
```diff
@@ -3944,11 +3944,11 @@ def wait_start(self, start_wait_sec):
-                ["bash", "-c", f"gdb -batch -ex 'thread apply all bt full' -p {pid}"],
-                user="root",
-            )
+        # if pid is not None:
+        #     self.exec_in_container(
+        #         ["bash", "-c", f"gdb -batch -ex 'thread apply all bt full' -p {pid}"],
+        #         user="root",
+        #     )
```
Why was this commented out?

tuanpach replied:
My bad, it took a too long time to dump threads when testing. I commented it out temporarily. However, I forget to revert it. 

I will revert it.
